### PR TITLE
add patch controller for change orderNumber

### PR DESCRIPTION
**Pull request by OleksandrOpanasiuk**

### CODE REVIEW

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 1]
```diff
@@ -175,4 +177,32 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
             throw new DayTodoReadyCanNotBeUpdatedException("Day todo ready can't be updated.", ex);
         }
     }
+
+    @PatchMapping
+    @ResponseStatus(HttpStatus.CREATED)
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: we don't create we update it here. OK status will be good

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 2]
```diff
@@ -175,4 +177,32 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+    @ApiOperation(value = "Change orderNumber id existed Todo", response = DayTodoResponseDto.class, responseContainer = "List")
+    @ApiResponses(value = {
+            @ApiResponse(code = 201, message = "Created"),
+            @ApiResponse(code = 400, message = "Invalid operation"),
+            @ApiResponse(code = 403, message = "Operation forbidden"),
+            @ApiResponse(code = 500, message = "Unexpected error")
+    })
+    public List<DayTodoResponseDto> changeOrderNumberDayTodo(@PathVariable Map<UUID,Integer> map ) {
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: you shouldn't name Map map, List list or Set set, it doesn't add meaning

Comment (FedosovMax) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Consider using another name, changeOrderNumberInDayTodo, changeDayTodoOrderNumber but it's up to you

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 3]
```diff
@@ -175,4 +177,32 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+            Map<UUID,Integer> mapExistDayTodo = map.entrySet().stream()
+                    .filter(entry -> dayTodoService.isExistDayTodo(entry.getKey()))
+                    .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+
+            List<DayTodoVO> dayTodoVOS = mapExistDayTodo.entrySet().stream()
+                    .map(entry ->dayTodoService.updateDayTodoById(entry.getKey(),entry.getValue()))
+                    .collect(Collectors.toList());
+
+            return dayTodoVOS.stream().map(dayTodoVO -> dayTodoRestMapper.toDayTodoResponseDto(dayTodoVO))
+                    .collect(Collectors.toList());
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Too much logic in the Controller, you should accept and validate data in a controller but all business logic and mappings should be done in Service. Maybe now it's not like this in other Controllers but we should make it like this.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 4]
```diff
@@ -175,4 +177,32 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+
+        } catch (RuntimeException ex) {
+            log.error("Order number does not changed.", ex);
+            throw new CreateDayTodoException("Order number does not changed.", ex);
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: You should use other exception not Create..

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java [hunk 5]
```diff
@@ -90,5 +90,17 @@ public DayTodoVO updateIsReady(UUID dayId, UUID dayTodoId, boolean isReady) {
         dayTodoVO = dayTodoMapper.toTodoVO(dayTodoRepository.save(dayTodoMapper.toTodo(dayTodoVO)));
         return experienceService.calculateExperience(dayTodoVO);
     }
+
+    @Transactional
+    public DayTodoVO updateDayTodoById(UUID dayTodoId, Integer newDayTodoOrderNumber) {
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: it should be orderNumber update?

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java [hunk 6]
```diff
@@ -90,5 +90,17 @@ public DayTodoVO updateIsReady(UUID dayId, UUID dayTodoId, boolean isReady) {
...
+        DayTodoVO dayTodoVO = findById(dayTodoId);
+        dayTodoVO.setOrderNumber(newDayTodoOrderNumber);
+        DayTodo dayTodo = dayTodoRepository.save(dayTodoMapper.toTodo(dayTodoVO));
+        return dayTodoMapper.toTodoVO(dayTodo);
+    }
+
+    public boolean isExistDayTodo(UUID dayTodoId) {
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: this method is redundant, you should have findById and throw exception 404 not found if it doesn't exist and not doing this additional boolean check

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java [hunk 7]
```diff
@@ -90,5 +90,17 @@ public DayTodoVO updateIsReady(UUID dayId, UUID dayTodoId, boolean isReady) {
...
+        return dayTodoRepository.findByIdAlive(dayTodoId).isPresent();
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: I don't see the point in adding this additional call to DB as well

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 1 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: I would like understand with statuses.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 2 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Your name is better.

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java (hunk 5 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: It's name better too.

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java (hunk 6 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: I thought that's validate it means update all exist dayTodo for synchronize back and front. If I throw exception we don't update any order number. Ok I will change this method.

File: todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java (hunk 7 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/service/DayTodoService.java]: Ok. I deleted this method.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 8]
```diff
@@ -175,4 +178,26 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+            @ApiResponse(code = 403, message = "Operation forbidden"),
+            @ApiResponse(code = 500, message = "Unexpected error")
+    })
+    public List<DayTodoResponseDto> changeDayTodoOrderNumber(@PathVariable Map<UUID,Integer> orderNumbersMap) {
+        try {
+            orderNumbersMap.keySet().stream().map(dayTodoService::findById);
+            List<DayTodoVO> dayTodoVOS = orderNumbersMap.entrySet().stream()
+                    .map(entry ->dayTodoService.orderNumberUpdate(entry.getKey(),entry.getValue()))
+                    .collect(Collectors.toList());
+            return dayTodoVOS.stream().map(dayTodoRestMapper::toDayTodoResponseDto).collect(Collectors.toList());
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: This should be done in Service. In Resource you should accept request, validate it with annotations in method params or in Dtos and send all information to service. d

File: todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package com.knighttodo.todocore.exception;
+
+public class UpdateOrderNumberException extends RuntimeException{
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java]: autoformat , no space before the last symbol in the line

File: todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+
+    public UpdateOrderNumberException(String message, Throwable cause) { super(message, cause);}
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java]: format

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 11]
```diff
@@ -175,4 +178,26 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
             throw new DayTodoReadyCanNotBeUpdatedException("Day todo ready can't be updated.", ex);
         }
     }
+
+    @PatchMapping("")
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: you don't need (""), just @PatchMapping

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 12]
```diff
@@ -175,4 +178,26 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+    @ResponseStatus(HttpStatus.OK)
+    @ApiOperation(value = "Change orderNumber id existed Todo", response = DayTodoResponseDto.class, responseContainer = "List")
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Change orderNumber id in existing Todo looks better, use Google translate or MS Word to check spelling, my browser checks it.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 13]
```diff
@@ -175,4 +178,26 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+    @ApiResponses(value = {
+            @ApiResponse(code = 201, message = "Created"),
+            @ApiResponse(code = 400, message = "Invalid operation"),
+            @ApiResponse(code = 403, message = "Operation forbidden"),
+            @ApiResponse(code = 500, message = "Unexpected error")
+    })
+    public List<DayTodoResponseDto> changeDayTodoOrderNumber(@RequestBody Map<UUID,Integer> orderNumbersMap) {
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: don't add to the name of the Map "Map" like to the List "List", it's redundant.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java [hunk 14]
```diff
@@ -175,4 +178,26 @@ public DayTodoReadyResponseDto updateIsReady(@PathVariable UUID dayId, @PathVari
...
+        try {
+            orderNumbersMap.keySet().stream().map(dayTodoService::findById);
```
Comment (FedosovMax) [inline todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Just send this Map to the service, in Controller you receive the request and send it to the service and return a response, everything else is redundant.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 12 above)
Comment (FedosovMax) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Use Grammarly plugin for the browser and you can use it online.

File: todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java [hunk 15]
```diff
@@ -331,5 +340,22 @@ public void findDayTodosByDayId_shouldReturnExistingTodo_whenIdIsCorrect() throw
                 .andExpect(status().isFound())
                 .andExpect(jsonPath(buildJsonPathToLength()).value(2));
     }
+    //my test
```
Comment (FedosovMax) [inline todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java]: redundant comment

File: todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java [hunk 16]
```diff
@@ -331,5 +340,22 @@ public void findDayTodosByDayId_shouldReturnExistingTodo_whenIdIsCorrect() throw
...
+        Day day = dayRepository.save(DayFactory.dayInstance());
+        DayTodo firstDayTodo = dayTodoRepository.save(DayTodoFactory.dayTodoWithDayInstance(day));
+        dayTodoRepository.save(DayTodoFactory.dayTodoWithDayInstance(day));
+
+        Map<UUID, Integer> map = new HashMap<>();
+        map.put(firstDayTodo.getId(), 777);
+
+        mockMvc.perform(patch(API_BASE_URL_V1 + API_BASE_DAYS + "/" + day.getId() + API_BASE_TODOS)
+                .content(objectMapper.writeValueAsString(map))
+                .contentType(MediaType.APPLICATION_JSON_VALUE))
```
Comment (FedosovMax) [inline todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java:343]: check please if you return the updated order number in json response

File: todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java (hunk 9 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java]: Done

File: todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java (hunk 10 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/exception/UpdateOrderNumberException.java]: Done

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 11 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: It's result my experiments

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 12 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Done.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 13 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Ok. It's result of learning.

File: todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java (hunk 15 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java]: Yes. Done.

File: todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java (hunk 14 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/main/java/com/knighttodo/todocore/rest/DayTodoResource.java]: Done.

File: todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java (hunk 16 above)
Comment (OleksandrOpanasiuk) [reply todocore/src/test/java/com/knighttodo/todocore/integration/DayTodoResourceIntegrationTest.java:343]: Done.

**Review by FedosovMax — APPROVED** [review]

**Pull request merged**